open-craft/ecommerce
# BB-4116: Add management command for hyperpay report

Adds a management command to report hyperpay manual review required ( see in [result codes](https://hyperpay.docs.oppwa.com/reference/resultCodes#successful)] type responses.

**Related Tickets**: [BB-4116](https://tasks.opencraft.com/browse/BB-4116)

### Testing Instructions:

1. Checkout to this branch
2. Get `ecommerce` running in `juniper` devstack (Preferable NELP devstack)
3. Attach to ecommerce using `make ecommerce-shell`

#### Unittest

1. Run unittests and verify that they pass (Use `./manage.py test extensions.payment.management.tests.test_hyperpay_report`)

#### Manual Test
1. Open shell using `./manage.py shell`
2. Create PaymentProcessorResponse objects by running the following code.
```
import datetime
from django.utils.timezone import now
from ecommerce.extensions.payment.models import PaymentProcessorResponse

PaymentProcessorResponse.objects.create(processor_name="hyperpay", response={"result": {"code": "000.400.000"}})
PaymentProcessorResponse.objects.create(processor_name="hyperpay_mada", response={"result": {"code": "000.400.000"}})
p1 = PaymentProcessorResponse.objects.create(processor_name="hyperpay", response={"result": {"code": "000.400.000"}})
p1.created = now() - datetime.timedelta(hours=9)
p1.save()
p2 = PaymentProcessorResponse.objects.create(processor_name="hyperpay_mada", response={"result": {"code": "000.400.000"}})
p2.created = now() - datetime.timedelta(hours=9)
p2.save()
```
3. exit shell
4. Run `./manage.py hyperpay_report --email email1@example.com email2@example.com`
5. Verify that email recipients contain email1@example.com email2@example.com
6. Verfiy that the email html has 2 table rows for bad responses
7.  Run `./manage.py hyperpay_report --duration 10 --email email1@example.com email2@example.com`
8. Verify that email recipients contain email1@example.com email2@example.com
6. Verfiy that the email html has 4 table rows for bad responses

Note: The email content will have lists with empty values as there are no Orders associated with them
